prov/usnic/src/usdf_cq.c

```diff
@@ -916,14 +1010,174 @@ usdf_cq_process_attr(struct fi_cq_attr *attr, struct usdf_domain *udp)
 	return 0;
 }
 
+static int usdf_cq_fd_set_nonblock(int fd)
+{
+	int flags;
+
+	flags = fcntl(fd, F_GETFL, 0);
+	if (flags == -1) {
+		USDF_WARN_SYS(CQ, "fcntl getfl failed[%d]\n", errno);
+		return -errno;
+	}
+
+	if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) < 0) {
+		USDF_WARN_SYS(CQ, "fcntl setfl failed[%d]\n", errno);
+		return -errno;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int usdf_cq_create_fd(struct usdf_cq *cq)
+{
+	int ret;
+
+	ret = usd_ib_cmd_create_comp_channel(cq->cq_domain->dom_dev,
+			&cq->object.fd);
+	if (ret) {
+		USDF_WARN_SYS(CQ, "failed to create comp channel\n");
+		return -FI_EINVAL;
+	}
+
+	USDF_DBG_SYS(CQ, "successfully created comp channel with fd %d\n",
+			cq->object.fd);
+
+	/* Going to need this assuming edge-triggered semantics.
+	 */
+	return usdf_cq_fd_set_nonblock(cq->object.fd);
+}
+
+int usdf_cq_trywait(struct fid *fcq)
+{
+	struct usdf_cq *cq;
+	uint64_t ev;
+	int empty;
+	int ret;
+
+	cq = cq_fidtou(fcq);
+
+	switch (cq->cq_attr.wait_obj) {
+	case FI_WAIT_UNSPEC:
+		return FI_SUCCESS;
+	case FI_WAIT_FD:
+	case FI_WAIT_SET:
+		break;
+	default:
+		USDF_WARN_SYS(CQ, "unsupported wait object type\n");
+		return -FI_EINVAL;
+	}
+
+	while (1) {
+		errno = 0;
+		ret = read(cq->object.fd, &ev, sizeof(ev));
+		if (ret == 0)
```

Comment:
If `(ret == 0)`, that means the other end of the fd closed, right?  Wouldn't that be an error?